crates/ntx-builder/src/transaction.rs

```diff
@@ -348,12 +386,70 @@ impl DataStore for NtxDataStore {
         }
     }
 
+    /// Retrieves a note script by its root hash.
+    ///
+    /// This implementation uses the configured RPC client to call the `GetNoteScriptByRoot`
+    /// endpoint on the RPC server.
     fn get_note_script(
         &self,
         script_root: Word,
-    ) -> impl FutureMaybeSend<Result<miden_objects::note::NoteScript, DataStoreError>> {
-        // TODO: Add implementation for getting note script from NtxDataStore.
-        async move { Err(DataStoreError::NoteScriptNotFound(script_root)) }
+    ) -> impl FutureMaybeSend<Result<NoteScript, DataStoreError>> {
+        let mut rpc_client = self.rpc_client.clone();
+        let cache = self.script_cache.clone();
+
+        async move {
+            // Attempt to retrieve the script from the cache.
+            if let Some(cached_script) = {
+                let mut cache_guard =
+                    cache.lock().expect("cache mutex cannot already be held in the current thread");
+                cache_guard.get(&script_root).cloned()
+            } {
+                return Ok(cached_script);
+            }
+
+            // Retrieve the script from the RPC.
+            let note_root = NoteRoot { root: Some(script_root.into()) };
+            match rpc_client.get_note_script_by_root(note_root).await {
+                Ok(response) => {
+                    // Decode the script from the response.
+                    let maybe_script = response.into_inner();
+                    if let Some(script_proto) = maybe_script.script {
+                        match miden_objects::note::NoteScript::read_from_bytes(&script_proto.mast) {
+                            Ok(script) => {
+                                // Cache the retrieved script.
+                                {
+                                    let mut cache_guard = cache.lock().unwrap();
+                                    cache_guard.put(script_root, script.clone());
+                                }
+                                // Return script.
+                                Ok(script)
+                            },
+                            // Failed to decode the script.
+                            Err(err) => {
+                                tracing::error!(
+                                    target: COMPONENT,
+                                    ?err,
+                                    "failed to deserialize note script from bytes"
+                                );
+                                Err(DataStoreError::NoteScriptNotFound(script_root))
```

Comment:
I'm not sure this should return `NoteScriptNotFound` - technically, it was found but was just malformed. Maybe some other error is more appropriate.